File: datasets/synthetic/SSRF-PY-REPO-002/src/services/proxy_service.py
import ipaddress
from urllib.parse import urlparse

from utils.http_utils import get_url

_ALLOWED_SCHEMES = {"https"}
_ALLOWED_DOMAINS = {"api.example.com", "cdn.example.com"}


class ProxyService:
    def _is_safe_url(self, url: str) -> bool:
        try:
            parsed = urlparse(url)
        except ValueError:
            return False

        if parsed.scheme not in _ALLOWED_SCHEMES:
            return False

        hostname = parsed.hostname
        if hostname is None or hostname not in _ALLOWED_DOMAINS:
            return False

        try:
            addr = ipaddress.ip_address(hostname)
            if addr.is_private or addr.is_loopback or addr.is_link_local:
                return False
        except ValueError:
            pass

        return True

    def fetch_remote(self, url: str) -> str | None:
        if not self._is_safe_url(url):
            return None
        return get_url(url)

File: datasets/synthetic/SSRF-JA-REPO-002/src/main/java/com/example/demo/service/ProxyService.java
package com.example.demo.service;

import com.example.demo.util.HttpUtil;

import java.net.InetAddress;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Set;

public class ProxyService {

    private static final Set<String> ALLOWED_SCHEMES = Set.of("https");
    private static final Set<String> ALLOWED_DOMAINS =
            Set.of("api.example.com", "cdn.example.com");

    private boolean isSafeUrl(String rawUrl) {
        try {
            URL parsed = new URL(rawUrl);
            if (!ALLOWED_SCHEMES.contains(parsed.getProtocol())) {
                return false;
            }
            String host = parsed.getHost();
            if (host == null || !ALLOWED_DOMAINS.contains(host)) {
                return false;
            }
            try {
                InetAddress addr = InetAddress.getByName(host);
                if (addr.isLoopbackAddress() || addr.isSiteLocalAddress()
                        || addr.isLinkLocalAddress()) {
                    return false;
                }
            } catch (Exception ignored) {
                // DNS解析失败时静默放行；DNS rebinding场景不在本模型范围内
            }
            return true;
        } catch (MalformedURLException e) {
            return false;
        }
    }

    public String fetchRemote(String url) {
        if (!isSafeUrl(url)) {
            return null;
        }
        return HttpUtil.getUrl(url);
    }
}
